cdc/sinkv2/eventsink/txn/txn_sink.go

```diff
@@ -14,24 +14,204 @@
+	"go.uber.org/zap"
+)
+
+const (
+	defaultConflictDetectorSlots int64        = 1024 * 1024
+	crcTable                     *crc64.Table = crc64.MakeTable(crc64.ISO)
 )
 
 // Assert EventSink[E event.TableEvent] implementation
-var _ eventsink.EventSink[*model.SingleTableTxn] = (*Sink)(nil)
+var _ eventsink.EventSink[*model.SingleTableTxn] = (*sink)(nil)
+
+// sink is the sink for SingleTableTxn.
+type sink struct {
+	conflictDetector *causality.ConflictDetector[*worker, *txnEvent]
+	workers          []*worker
+}
 
-// Sink is the sink for SingleTableTxn.
-type Sink struct{}
+func newSink(backends []backend, conflictDetectorSlots int64) sink {
+	workers := make([]*worker, 0, len(backends))
+	for i, backend := range backends {
+		w := newWorker(i, backend)
+		w.wg.Add(1)
+		go w.Run()
+		workers = append(workers, w)
+	}
+	detector := causality.NewConflictDetector[*worker, *txnEvent](workers, conflictDetectorSlots)
+	return sink{detector, workers}
+}
 
 // WriteEvents writes events to the sink.
-func (s *Sink) WriteEvents(rows ...*eventsink.TxnCallbackableEvent) {
-	// TODO implement me
-	panic("implement me")
+func (s *sink) WriteEvents(rows ...*eventsink.TxnCallbackableEvent) {
+	for _, row := range rows {
+		_ = s.conflictDetector.Add(&txnEvent{row, nil})
+	}
+}
+
+// Close closes the sink. It won't wait for all pending items backend handled.
+func (s *sink) Close() error {
+	for _, w := range s.workers {
+		w.Close()
+	}
+	s.conflictDetector.Close()
+	return nil
+}
+
+type txnEvent struct {
+	*eventsink.TxnCallbackableEvent
+	conflictKeys []int64
+}
+
+// ConflictKeys implements causality.txnEvent interface.
+func (e *txnEvent) ConflictKeys() (sums []int64) {
+	if len(e.conflictKeys) > 0 {
+		return e.conflictKeys
+	}
+
+	keys := genTxnKeys(e.TxnCallbackableEvent.Event)
+	sums = make([]int64, 0, len(keys))
+	for _, key := range keys {
+		hasher := crc64.New(crcTable)
+		if _, err := hasher.Write(key); err != nil {
+			log.Panic("crc64 hasher fail")
+		}
+		sums = append(sums, int64(hasher.Sum64()))
+	}
+	e.conflictKeys = sums
+	return
+}
+
+type worker struct {
+	ID      int
+	txnCh   *chann.Chann[txnWithNotifier]
+	stopped chan struct{}
+	wg      sync.WaitGroup
+
+	backend backend
+}
+
+type txnWithNotifier struct {
+	*txnEvent
+	wantMore func()
+}
+
+func newWorker(ID int, backend backend) *worker {
+	return &worker{
+		ID:      ID,
+		txnCh:   chann.New[txnWithNotifier](chann.Cap(-1)),
+		stopped: make(chan struct{}),
+		backend: backend,
+	}
+}
+
+func (w *worker) Add(txn *txnEvent, unlock func()) {
+	w.txnCh.In() <- txnWithNotifier{txn, unlock}
+}
+
+func (w *worker) Close() {
+	close(w.stopped)
+	w.wg.Wait()
+}
+
+func (w *worker) Run() {
+	defer w.wg.Done()
+	for {
+		select {
+		case <-w.stopped:
+			return
+		case txn := <-w.txnCh.Out():
+			txn.wantMore()
+			if err := w.backend.onTxnEvent(txn.txnEvent.TxnCallbackableEvent); err != nil {
+				// TODO: handle errors.
+				return
+			}
+		case <-w.backend.timer().C:
```

Comment:
It returns a pointer. The backend implementation can decide whether to create a new one or not. Generally it should always return one same object.

`Ticker` is not suite here because `backend.onTxnEvent` may flush pending items internally, in which case the timer should be reset.